### ISSUE TITLE
Running Xournal++ on reMarkable tablet

### ISSUE THREAD

**Issue by giampierosalvi** [xournalpp/xournalpp#2018]
It would be great to be able to run Xournal++ on the reMarkable tablet for (at least) two reasons:
- being able to zoom in and out more easily when annotating PDF files (the screen is too small to display a full A4 document, and the resizing capabilities are very limited in the tablet),
- being able to move .xopp files back and forward between tablet and desktop to work on both platforms.

Do you know how much work this would be? The tablet runs a modified Linux kernel: 4.9.84-zero-gravitas, on an 32 bits ARM processor: armv7l. It runs xochitl as UI which is built on top of QT for what I can tell. I believe most relevant libraries are accessible. Not sure how easy it is to install Xournal++'s dependencies.

An alternative would be to install a full linux distribution on the tablet, but that's quite overkill. Also the distribution would probably drain the battery much faster, and most desktop interfaces are not designed for e-ink screens.

If you need more information from the device, I can add it. There is a lot of documentation, for example at the following web sites:
https://remarkablewiki.com/devel/start
https://github.com/reHackable/awesome-reMarkable

Thank you!

**Comment by Technius** [comment]
I don't think anyone has tried compiling with ARM yet, but the main blocker is that we aren't sure if 32-bit architectures are supported (#1747).

**Comment by LittleHuba** [comment]
I have no idea if Xournal++ would run on ARM. This depends most likely on the support by our dependencies. We don't have any specific code that is fixed to any architecture. 

The problem with 32bit is that we highly rely on `size_t` which differs in width on 64bit and 32bit. While I don't think we do have any arithmetic that would break there might be some line of code that does take offence of this change which I haven't found yet. 
Somebody tried once to add 32bit support but stumbled on some kind of problem with GTK. Don't remember quite what it was. He was unable to fix it back then because he never found the cause of it. 

There is no other way than just trying your hand at building it on your device. Set up a toolchain and try your hand. If you run into trouble we can give you a hand.

**Comment by dkwo** [comment]
This is of course a great idea.
Another variation could be to use the RM2 tablet as a sort of input for xournal++ on the PC, such that while drawing on the eink screen the output also appears on the PC: does this make sense?
I do not know if running full xouran++ on rm2 is needed for this, or perhaps it can be done in another way..

**Comment by AlessandroW** [comment]
I had the same thought! 
Another approach (or a small step in that direction) would be to convert the rm files to xournal++ format. Do you know if that's feasible?

**Comment by LittleHuba** [comment]
I have no idea what the rm files look like but Xournal++ uses a simple XML format that is compressed using `gzip`. So as long as you can read the rm format you'll most likely be able to translate it to our format (at least to some extent).

**Comment by AlessandroW** [comment]
I'll try to extract a file and figure this out. :+1:

**Comment by AlessandroW** [comment]
@LittleHuba is this somewhat helpful: https://github.com/lschwetlick/maxio/blob/master/rm_tools/reverse_engineering/convert23.ipynb ?

**Comment by LittleHuba** [comment]
Yes this goes in the right direction. This is more or less the first half of the whole job. Now there is only the part missing where you convert this known format to the Xournal++ format.

**Comment by AlessandroW** [comment]
Awesome. 
Can you point me towards a resource for understanding the Xournal++ file format? That'd be helpful. 
Thank you!

**Comment by LittleHuba** [comment]
There is unfortunately currently no resource for this. You may create a simple document with some different elements in Xournal++ and then just look at the resulting file. Be aware that the file is compressed using gzip. After you uncompress it, the content is simple XML which is very easy to read and understand.

**Comment by rolandlo** [comment]
The .xopp file format is still pretty close to Xournal's .xoj-file format. The latter is documented [here](http://xournal.sourceforge.net/manual.html#file-format). You will see some differences between the two file formats in the `XojExportHandler` class (which is derived from the `SaveHandler` class and removes all features that break compatibility with Xournal).

**Comment by giampierosalvi** [comment]
A while back, I made a tool to convert between the xournal++ format and the format used by the Write app which is a combination of single page svg and html to put the pages together. This is by no means a general solution, and it is closer to a hack for what I needed at the time, but it could be useful to start digging into the xournal++ format. Hope it's useful

https://github.com/giampierosalvi/xournal-tools

**Comment by Moelf** [comment]
I came across a Reddit [post](https://www.reddit.com/r/RemarkableTablet/comments/gkktxy/desktop_linux_on_remarkable_xournal_doom_and_more/) that shows running a proper Linux AND Xournal?

**Comment by jopereira** [comment]
As suggested in previous comments, I adapted rM2svg from maxio to generate a .xopp file: https://github.com/jopereira/rm2xopp

**Comment by mahrud** [comment]
> As suggested in previous comments, I adapted rM2svg from maxio to generate a .xopp file: https://github.com/jopereira/rm2xopp

Is there any instructions for how to use it? I'm not sure what the input format is supposed to be.

**Comment by jopereira** [comment]
I added minimal instructions to the README file (and fixed a bug).

**Comment by jcjgraf** [comment]
I have got a remarkable 1 (RM1) and I have installed Arch ARM on it. I was able to build all dependencies as well as xournalpp. Unfortunately, the performance is really bad. Xournalpp is not just lagging while drawing (0.5 - 1.0s delay) but it also often does not draw correctly. I.e. when I draw something U-shaped, it may simply draw a line between the start and endpoint. In comparison, Xournal and MrWrite lag too, though much less (0.1 - 0.3s). And in addition, they draw lines accurately.

I am not sure it is due to the really underspecified hardware, or rather due some "32-bit bug". I have tried @Lukey3332's 32bit version posed in #1747, but the behaviour is the same.

I have not tested any more "advanced" features (pdf reading, highlighting, images etc.).

It would be absolutely awesome to be able to run Xorunalpp on low-spec (e-ink) devices like the RM. Especially, since there are no real alternative software available.

**Comment by dkwo** [comment]
as far as arm goes, there should be no issue: xournalpp is packaged in alpine and void linux, so also musl should be fine. the hardware itself may be too crappy though. btw, i stumbled upon the pine64 pinenote, which has much better specs.

**Comment by RohanModi-CA** [comment]
> As suggested in previous comments, I adapted rM2svg from maxio to generate a .xopp file: https://github.com/jopereira/rm2xopp

Out of necessity, I [threw something similar together](https://github.com/RohanModi-CA/rm-xopp-converter) for the rm2, for my own use case but feel free to take a look if anyone's interested. I have it go both ways, which is the issue I had with that script but that seems much more developed; I only translate pen strokes. Would need a lot more work to be a complete product, but it works for my use case.
